lib/figures/german.rb

```diff
@@ -1,139 +1,136 @@
+        ten(ten_digit)
+      ].join
 
-      @number < 0 ? "minus #{result}" : result
+      word = append_exponent_identifier(word, triple_index)
+      cleanup_exceptions(word)
     end
 
-    def write_triple(triple, triple_index)
-      hundred, ten, one = triple.match(/\A(\d)??(\d)??(\d)\z/).captures.map(&:to_i)
-      result = ""
+    # splits up a triple into hundreds, tens and unit position
+    def split_triple(triple)
+      triple.match(/\A(\d)??(\d)??(\d)\z/).captures.map(&:to_i)
+    end
+
+    # returns the word for the given unit number
+    def unit(digit)
+      return '' if digit.zero?
+      UNITS[digit - 1]
+    end
 
-      is_empty    = hundred == 0 && ten == 0 && one == 0
-      is_thousand = !is_empty && triple_index == 1
+    # returns the copula between unit position and tens
+    def copula(unit_digit, ten_digit)
+      'und' if ten_digit > 1 && !unit_digit.zero?
+    end
 
-      if hundred && hundred != 0
-        result << WORDS[:first10][hundred] + WORDS[:hundred]
+    # returns the word for the given tens digit
+    def ten(digit)
+      case digit
+      when 0 then ''
+      when 1 then 'zehn'
+      else unit(digit) + 'zig'
       end
+    end
 
-      if !ten
-        result << WORDS[:first10][one]
-      elsif ten == 0
-        if !is_thousand && hundred && one == 1
-          if triple_index > 1
-            result << WORDS[:one]
-          else
-            result << WORDS[:single]
-          end
-        else
-          result << WORDS[:first10][one]
-        end
-      elsif ten == 1
-        result << WORDS[:second10][one]
-      else
-        result << WORDS[:first10][one]
-        result << WORDS[:copula] unless one == 0
-        result << WORDS[:tens][ten - 2]
+    # returns the word for the given hundreds number
+    def hundred(digit)
+      case digit
+      when 0 then ''
+      else unit(digit) + 'hundert'
       end
+    end
+
+    # adds the exponent word to the triple word
+    # e.g. tausend for the second triple   (index = 1)
+    #      Million for the third triple    (index = 2)
+    #      Milliarde for the fourth triple (index = 3)
+    #
+    #      indexes => PREFIXES index
+    #      2,3 => 1; 4,5 => 2; 6,7 => 3; ... : floored division by 2
+    # etc.
+    def append_exponent_identifier(word, index)
+      return word if word.empty? || index.zero? || triples_count == 1
 
-      if is_thousand
-        result << WORDS[:thousand]
-      elsif triple_index > 1 && !is_empty
-        huge = WORDS[:huge][triple_index - 2].dup
-        huge[/e?$/] = "en" unless one == 1
-        result << " #{huge} "
+      if index == 1
+        word + PREFIXES[:units][0]
+      elsif index.even?
+        pluralize(word + ' ' + (PREFIXES[:units][index / 2] + "llion ").capitalize)
+      elsif index.odd?
+        pluralize(word + ' ' + (PREFIXES[:units][index / 2] + "lliarde ").capitalize)
       end
+    end
+
+    # pluralizes exponent identifiers
+    def pluralize(word)
```

Comment:
@janlelis I'm not sure whether `pluralize` is the right name here, because the method handles also decision logic whether to apply a plurarilzation (so it is more a " maybe pluralize"). Maybe we can still improve that a bit.